File: app/src/main/java/com/chooloo/www/callmanager/ui/cursor/CursorFragment.java
package com.chooloo.www.callmanager.ui.cursor;

import android.database.Cursor;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;
import androidx.recyclerview.widget.RecyclerView;

import com.chooloo.www.callmanager.R;
import com.chooloo.www.callmanager.databinding.FragmentItemsBinding;
import com.chooloo.www.callmanager.ui.base.BaseFragment;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.chooloo.www.callmanager.util.AnimationUtils.runLayoutAnimation;

public abstract class CursorFragment<A extends CursorAdapter> extends BaseFragment implements CursorMvpView, LoaderManager.LoaderCallbacks<Cursor> {


    protected static int mLoaderId = 1;

    protected A mAdapter;
    private CursorMvpPresenter<CursorMvpView> mPresenter;
    protected FragmentItemsBinding binding;
    private RecyclerView.OnScrollListener mOnScrollListener;

    public CursorFragment() {
        mOnScrollListener = new RecyclerView.OnScrollListener() {
        };
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentItemsBinding.inflate(inflater);
        return binding.getRoot();
    }

    @NonNull
    @Override
    public Loader<Cursor> onCreateLoader(int id, @Nullable Bundle args) {
        return mPresenter.onCreateLoader(id, args);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor data) {
        mPresenter.onLoadFinished(loader, data);
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> loader) {
        mPresenter.onLoaderReset(loader);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detach();
    }

    @Override
    public void onSetup() {
        mPresenter = new CursorPresenter<>();
        mPresenter.attach(this);

        mAdapter = onGetAdapter();

        binding.refreshLayout.setOnRefreshListener(() -> mPresenter.onRefresh());
        binding.itemsRecyclerView.setAdapter(mAdapter);
        binding.itemsRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                mOnScrollListener.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                mOnScrollListener.onScrolled(recyclerView, dx, dy);
            }
        });

        // TODO make this no permission handling better
        if (!hasPermissions()) {
            mPresenter.onNoPermissions();
            binding.emptyState.emptyTitle.setText(getString(R.string.empty_list_no_permissions));
        }
    }

    @Override
    public void updateData(Cursor cursor) {
        mAdapter.setCursor(cursor);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mAdapter.getItemCount();
    }

    @Override
    public void load() {
        if (hasPermissions()) {
            runLoader();
        } else if (mPresenter != null) {
            mPresenter.onNoPermissions();
        }
    }

    @Override
    public void runLoader() {
        if (LoaderManager.getInstance(this).getLoader(mLoaderId) == null) {
            LoaderManager.getInstance(this).initLoader(mLoaderId, getArgsSafely(), this);
        } else {
            LoaderManager.getInstance(this).restartLoader(mLoaderId, getArgsSafely(), this);
        }
    }

    @Override
    public void showEmptyPage(boolean isShow) {
        binding.emptyState.emptyState.setVisibility(isShow ? VISIBLE : GONE);
        binding.itemsRecyclerView.setVisibility(isShow ? GONE : VISIBLE);
    }

    @Override
    public void animateListView() {
        runLayoutAnimation(binding.itemsRecyclerView);
    }

    @Override
    public void setRefreshing(boolean isRefresh) {
        if (!isRefresh && binding.refreshLayout.isRefreshing()) {
            binding.refreshLayout.setRefreshing(false);
        } else if (isRefresh && !binding.refreshLayout.isRefreshing()) {
            binding.refreshLayout.setRefreshing(true);
        }
    }

    @Override
    public void setOnScrollListener(RecyclerView.OnScrollListener onScrollListener) {
        mOnScrollListener = onScrollListener;
    }

    public abstract A onGetAdapter();
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/cursor/CursorFragment.kt
package com.chooloo.www.callmanager.ui.cursor

import android.Manifest.permission
import android.database.Cursor
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.chooloo.www.callmanager.R
import com.chooloo.www.callmanager.adapter.CursorAdapter
import com.chooloo.www.callmanager.databinding.FragmentItemsBinding
import com.chooloo.www.callmanager.ui.base.BaseFragment
import com.chooloo.www.callmanager.util.AnimationUtils.runLayoutAnimation

abstract class CursorFragment<A : CursorAdapter<*>> : BaseFragment(), CursorMvpView {

    private var _onScrollListener: RecyclerView.OnScrollListener? = null
    private lateinit var _presenter: CursorMvpPresenter<CursorMvpView>
    protected lateinit var adapter: A
    protected lateinit var binding: FragmentItemsBinding

    override val itemCount: Int
        get() = adapter.itemCount

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentItemsBinding.inflate(inflater)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _presenter.detach()
    }

    override fun onSetup() {
        _presenter = CursorPresenter()
        _presenter.attach(this)

        adapter = onGetAdapter()

        binding.run {
            itemsRecyclerView.adapter = adapter
            itemsRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    _onScrollListener?.onScrollStateChanged(recyclerView, newState)
                }

                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    _onScrollListener?.onScrolled(recyclerView, dx, dy)
                }
            })
        }

        // TODO make this no permission handling better
        if (!hasPermission(permission.READ_CONTACTS)) {
            _presenter.onNoPermissions()
        }
    }

    override fun updateData(cursor: Cursor) {
        adapter.setCursor(cursor)
        adapter.notifyDataSetChanged()
    }

    override fun showEmptyPage(isShow: Boolean) {
        binding.emptyState.emptyTitle.text = getString(R.string.empty_list_no_results)
        binding.emptyState.emptyState.visibility = if (isShow) View.VISIBLE else View.GONE
        binding.itemsRecyclerView.visibility = if (isShow) View.GONE else View.VISIBLE
    }

    override fun showNoPermissions(isShow: Boolean) {
        binding.emptyState.emptyTitle.text = getString(R.string.empty_list_no_permissions)
        binding.emptyState.emptyState.visibility = if (isShow) View.VISIBLE else View.GONE
        binding.itemsRecyclerView.visibility = if (isShow) View.GONE else View.VISIBLE
    }

    override fun animateListView() {
        runLayoutAnimation(binding.itemsRecyclerView)
    }

    override fun setOnScrollListener(onScrollListener: RecyclerView.OnScrollListener) {
        _onScrollListener = onScrollListener
    }

    abstract fun onGetAdapter(): A
}